docs/docs/reference/release-notes/v3.0/index.md

```diff
@@ -0,0 +1,271 @@
+- [webpack 5](#webpack-5)
+- [React 17](#react-17)
+- [GraphQL 15](#graphql-15)
+- [ESLint 7](#eslint-7)
+
+Also check out [notable bugfixes and improvements](#notable-bugfixes-and-improvements).
+
+**Bleeding Edge:** Want to try new features as soon as possible? Install `gatsby@next` and let us know
+if you have any [issues](https://github.com/gatsbyjs/gatsby/issues).
+
+[Previous release notes](/docs/reference/release-notes/v2.32)
+
+[Full changelog](https://github.com/gatsbyjs/gatsby/compare/gatsby@2.32.0-next.0...gatsby@3.0.0)
+
+## Breaking Changes
+
+If you're looking for an overview of all breaking changes and how to migrate, please see the [migrating from v2 to v3 guide](/docs/reference/release-notes/migrating-from-v2-to-v3/).
+
+## Incremental Builds in OSS
+
+Gatsby v2 introduced experimental "Conditional Page Builds" (enabled by `GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES` environment variable). It had some gotchas and quirks and wasn't ready yet for GA. With Gatsby v3 we improved this feature, and activated it by default for everyone! So incremental builds is available in OSS now. This improvement is (re)generating only subset of HTML files that needs to be generated. To be able to use it you will need to keep to `.cache` and `public` directories from previous builds.
+
+Take a project powered by Shopify as an example. You have your listing of all products and then individual product pages -- when you change one single product, only that page should be rebuilt. In the screenshot below you can see exactly that (the sentence "Hello World" was added to the description):
+
+![Side-by-side view of a Shopify store instance on the left, and the preview on the right. At the bottom the terminal shows that after the change only one page was rebuilt](./incremental-builds-in-oss.jpg)
+
+The screenshot is taken from a lengthy video about Gatsby v3 at [GatsbyConf](https://gatsbyconf.com/). You can view the video showcasing this feature on YouTube there.
+
+### How does it work?
+
+Gatsby tracks "inputs" when generating HTML files. When those inputs change since the last build, the HTML files are marked to be regenerated (if they don't change we can reuse HTML files generated in previous build). In particular we track:
+
+- which page template the page is using
+- result of page query
+- results of static queries used by a page template
+- frontend source code (shared and also browser `gatsby-browser` / SSR `gatsby-ssr` specifically)
+
+### Gotchas
+
+As we mentioned, Gatsby tracks "inputs" used to generate HTML files. However, the `gatsby-ssr` file allows some arbitrary code execution. This includes for example `fs` reads. While Gatsby could also track files that are read, the custom code that does those reads might have some special logic that Gatsby is not aware of. If Gatsby discovers that those are used, it will disable "Incremental Builds"-mode to stay on the safe side (there will be warnings mentioning "unsafe builtin method").
```

Comment:
This paragraph is confusing to me. I'm not sure what the general takeaway is supposed to be.

Is it that Gatsby doesn't look at files that your code might be importing?

> If Gatsby discovers that **those** are used

What does "those" refer to?

Reply:
Imports are fine - in fact this is the migration path ( https://github.com/gatsbyjs/gatsby/blob/docs/v3-migration-guide/docs/docs/reference/release-notes/migrating-from-v2-to-v3.md#using-fs-in-ssr ). But in `gatsby-ssr.js` files, users (and plugins) are able to do:
```
const fs = require(`fs`)

const someUntrackedInput = fs.readFileSync(`some-path.txt`)
```

Gatsby will know that `fs.readFileSync` is used (and can know params it was called with), but it doesn't know context around it - would this be always the same file read? Or maybe the filename of the file that is read is generated and can change between builds. So because it doesn't exactly know intention and mechanics surrounding that "arbitrary" file reading (and how to safely make sure to regenerate needed html files in subsequent builds), it will want to stay on safe side and just generate all html files when something like that happens.

As for how to capture this so it's understandable ... I have no clue

> What does "those" refer to?

I mean using anything from builtin `fs` module ( https://nodejs.org/api/fs.html )

Reply:
@LekoArts updated the section in https://github.com/gatsbyjs/gatsby/pull/29910

@meganesu let us know if this is any better or it's still (very) confusing